src/openms_gui/source/VISUAL/TOPPViewTopDownViewBehavior.cpp

```diff
@@ -0,0 +1,1345 @@
+      cout << "Generated html:\n" << html << endl;
+    #endif
+    return html;
+  }
+
+
+  // add specializations to allow template implementation outside of header file:
+  template String TOPPViewTopDownViewBehavior::generateSequenceDiagram_<AASequence>(const AASequence& seq, const vector<PeptideHit::PeakAnnotation>& annotations, const StringList& top_ions, const StringList& bottom_ions);
+  template String TOPPViewTopDownViewBehavior::generateSequenceDiagram_<NASequence>(const NASequence& seq, const vector<PeptideHit::PeakAnnotation>& annotations, const StringList& top_ions, const StringList& bottom_ions);
+
+
+  void TOPPViewTopDownViewBehavior::addPrecursorLabels1D_(const vector<Precursor>& pcs)
+  {
+    LayerData& current_layer = tv_->getActive1DWidget()->canvas()->getCurrentLayer();
+
+    if (current_layer.type == LayerData::DT_PEAK)
+    {
+      const SpectrumType& spectrum = current_layer.getCurrentSpectrum();
+
+      for (vector<Precursor>::const_iterator it = pcs.begin(); it != pcs.end(); ++it)
+      {
+        // determine start and stop of isolation window
+        double center_mz = it->metaValueExists("isolation window target m/z") ?
+          double(it->getMetaValue("isolation window target m/z")) : it->getMZ();
+        double isolation_window_lower_mz = center_mz - it->getIsolationWindowLowerOffset();
+        double isolation_window_upper_mz = center_mz + it->getIsolationWindowUpperOffset();
+
+        // determine maximum peak intensity in isolation window
+        SpectrumType::const_iterator vbegin = spectrum.MZBegin(isolation_window_lower_mz);
+        SpectrumType::const_iterator vend = spectrum.MZEnd(isolation_window_upper_mz);
+
+        double max_intensity = (numeric_limits<double>::min)();
+        for (; vbegin != vend; ++vbegin)
+        {
+          if (vbegin->getIntensity() > max_intensity)
+          {
+            max_intensity = vbegin->getIntensity();
+          }
+        }
+
+        // DPosition<2> precursor_position = DPosition<2>(it->getMZ(), max_intensity);
+        DPosition<2> lower_position = DPosition<2>(isolation_window_lower_mz, max_intensity);
+        DPosition<2> upper_position = DPosition<2>(isolation_window_upper_mz, max_intensity);
+
+        Annotation1DDistanceItem* item = new Annotation1DDistanceItem(QString::number(it->getCharge()), lower_position, upper_position);
+        // add additional tick at precursor target position (e.g. to show if isolation window is asymmetric)
+        vector<double> ticks;
+        ticks.push_back(it->getMZ());
+        item->setTicks(ticks);
+        item->setSelected(false);
+
+        temporary_annotations_.push_back(item); // for removal (no ownership)
+        current_layer.getCurrentAnnotations().push_front(item); // for visualization (ownership)
+      }
+    }
+    else if (current_layer.type == LayerData::DT_CHROMATOGRAM)
+    {
+
+    }
+  }
+
+  void TOPPViewTopDownViewBehavior::removeTemporaryAnnotations_(Size spectrum_index)
```

Comment:
this method of adding temporary annotations is 1) not very intuitive 2) not well encapsulated 3) duplicated code

This could be moved into a separate class and added as member to LayerData

Reply:
hmm I think the temporary annotations are more related to a specific spectrum than a full layer. I would assume that showing a new spectrum or calling a CLASSX.removeTemporaryAnnotations() automatically removes all of them. Currently, Peak annotations are stored in LayerData. Would moving this to the Spectrum1DCanvas make more sense? LayerData doesn't seem a perfect fit.

Reply:
but I can move it there for now. removes the duplication